pool/src/graph.rs

```diff
@@ -179,6 +180,31 @@ impl DirectedGraph {
 		}
 	}
 
+	/// Promote any non-root vertices to roots based on current edges.
+	/// For a given tx, if there are no edges with that tx as destination then it is a root.
+	pub fn update_roots(&mut self) {
+		let mut new_roots = vec![];
+		for x in &self.vertices {
+			let edges = &self.edges.values().collect::<Vec<_>>();
+			match edges.iter().find(|edge| edge.destination == Some(x.transaction_hash)) {
+				Some(_) => (),
+				None => new_roots.push(x.clone()),
+			}
+		}
+
+		for x in new_roots {
+			self.roots.push(x.clone());
+		}
+
+		let mut new_vertices = vec![];
```

Comment:
I think you can avoid allocating a new vertices vector here and also avoid comparing against all of `roots` by reusing `new_roots`. Something like (not tested):

```
		for x in new_roots {
			self.roots.push(x.clone());
                        match self.vertices.iter().position(|v| v.transaction_hash == x.transaction_hash) {
                                   Some(i) => self.vertices.swap_remove(i),
                                    None=> (),
                         }
                  }
```

Even better, since you know where vertices are matching from the above loop (L187) you could do without this second position search altogether- although the ways I can think of to do that now are messy (would be really nice with Rust nightly's `drain_filter`, which isn't stable yet). That feels like an optimization that can wait, though.